File: packs/dsa-trees-ts/challenges/04-path-sum.test.ts
import { join } from 'node:path';
import { writeChallengeToTmp, importModule } from '../../test-helpers.js';

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val = 0, left: TreeNode | null = null, right: TreeNode | null = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function buildTree(values: (number | null)[]): TreeNode | null {
  if (!values.length || values[0] === null) return null;
  const root = new TreeNode(values[0]);
  const queue: TreeNode[] = [root];
  let i = 1;
  while (queue.length && i < values.length) {
    const node = queue.shift()!;
    if (i < values.length && values[i] !== null) {
      node.left = new TreeNode(values[i] as number);
      queue.push(node.left);
    }
    i++;
    if (i < values.length && values[i] !== null) {
      node.right = new TreeNode(values[i] as number);
      queue.push(node.right);
    }
    i++;
  }
  return root;
}

function treeToList(root: TreeNode | null): (number | null)[] {
  if (!root) return [];
  const result: (number | null)[] = [];
  const queue: (TreeNode | null)[] = [root];
  while (queue.length) {
    const node = queue.shift()!;
    if (node) {
      result.push(node.val);
      queue.push(node.left);
      queue.push(node.right);
    } else {
      result.push(null);
    }
  }
  while (result.length && result[result.length - 1] === null) result.pop();
  return result;
}

const CHALLENGE = join(import.meta.dirname, '04-path-sum.json');

describe('Path Sum', () => {
  let hasPathSum: (root: TreeNode | null, targetSum: number) => boolean;
  let cleanup: () => void;

  beforeAll(async () => {
    const tmp = writeChallengeToTmp(CHALLENGE);
    cleanup = tmp.cleanup;
    const mod = await importModule<{ hasPathSum: typeof hasPathSum }>(
      join(tmp.tmpDir, 'solution.ts')
    );
    hasPathSum = mod.hasPathSum;
  });

  afterAll(() => cleanup());

  it('returns true when path exists', () => {
    expect(
      hasPathSum(buildTree([5, 4, 8, 11, null, 13, 4, 7, 2, null, null, null, 1]), 22)
    ).toBe(true);
  });

  it('returns false when no path matches', () => {
    expect(hasPathSum(buildTree([1, 2, 3]), 5)).toBe(false);
  });

  it('returns false for empty tree', () => {
    expect(hasPathSum(null, 0)).toBe(false);
  });

  it('returns true for single node match', () => {
    expect(hasPathSum(buildTree([5]), 5)).toBe(true);
  });

  it('returns false for single node no match', () => {
    expect(hasPathSum(buildTree([5]), 1)).toBe(false);
  });

  it('returns true for left path', () => {
    expect(hasPathSum(buildTree([1, 2, 3]), 3)).toBe(true);
  });
});

File: packs/dsa-trees-py/challenges/04-path-sum.test.py
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def build_tree(values):
    """Build binary tree from level-order list. None means no node."""
    if not values or values[0] is None:
        return None
    root = TreeNode(values[0])
    queue = deque([root])
    i = 1
    while queue and i < len(values):
        node = queue.popleft()
        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1
        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1
    return root


def tree_to_list(root):
    """Convert tree to level-order list (with None for missing nodes)."""
    if not root:
        return []
    result = []
    queue = deque([root])
    while queue:
        node = queue.popleft()
        if node:
            result.append(node.val)
            queue.append(node.left)
            queue.append(node.right)
        else:
            result.append(None)
    while result and result[-1] is None:
        result.pop()
    return result


from solution import has_path_sum


def test_path_exists():
    root = build_tree([5, 4, 8, 11, None, 13, 4, 7, 2, None, None, None, 1])
    assert has_path_sum(root, 22) is True


def test_no_path():
    root = build_tree([1, 2, 3])
    assert has_path_sum(root, 5) is False


def test_empty_tree():
    assert has_path_sum(None, 0) is False


def test_single_node_match():
    root = build_tree([5])
    assert has_path_sum(root, 5) is True


def test_single_node_no_match():
    root = build_tree([5])
    assert has_path_sum(root, 1) is False


def test_left_path():
    root = build_tree([1, 2, 3])
    assert has_path_sum(root, 3) is True
